src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/validation/HaFlowValidationFsm.java

```diff
@@ -0,0 +1,334 @@
+        super(NEXT, ERROR, commandContext, carrier, eventListeners);
+        this.flowId = flowId;
+
+        service = new HaFlowValidationService(persistenceManager, new RuleManagerImpl(ruleManagerConfig));
+    }
+
+    protected void receiveData(State from, State to,
+                               Event event, Object context) {
+        HaFlow haFlow;
+        try {
+            haFlow = service.checkFlowStatusAndGetFlow(flowId);
+        } catch (FlowNotFoundException e) {
+            log.error("HaFlow {} not found when sending commands to SpeakerWorkerBolt", flowId, e);
+            sendException(e.getMessage(), "Receiving rules operation in HaFlowValidationFsm", ErrorType.NOT_FOUND);
+            return;
+        } catch (IllegalFlowStateException e) {
+            log.error("Could not validate flow: HaFlow {} is in DOWN state", flowId, e);
+            sendException("Could not validate flow",
+                    format("Could not validate flow: HaFlow %s is in DOWN state", flowId),
+                    ErrorType.UNPROCESSABLE_REQUEST);
+            return;
+        }
+
+        switchIdSet.addAll(extractSwitchIdSet(haFlow));
+
+        awaitingRules = switchIdSet.size();
+        log.debug("Send commands to get rules on the switches");
+        switchIdSet.forEach(switchId ->
+                getCarrier().sendSpeakerRequest(flowId, new DumpRulesForFlowHsRequest(switchId)));
+
+        Set<SwitchId> meterSwitchIds = new HashSet<>(haFlow.getEndpointSwitchIds());
+        meterSwitchIds.add(haFlow.getForwardPath().getYPointSwitchId());
+        Optional.ofNullable(haFlow.getProtectedForwardPath()).ifPresent(e -> meterSwitchIds.add(e.getYPointSwitchId()));
+
+        log.debug("Send commands to get meters on the termination switches");
+        awaitingMeters = meterSwitchIds.size();
+        meterSwitchIds.forEach(switchId -> {
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(switchId));
+        });
+
+        log.debug("Send commands to get groups on the termination switches");
+        awaitingGroups = 2;
+
+        getCarrier().sendSpeakerRequest(flowId,
+                new DumpGroupsForFlowHsRequest(haFlow.getForwardPath().getYPointSwitchId()));
+        Optional.ofNullable(haFlow.getProtectedForwardPath()).ifPresent(protectedFlowPath -> {
+            getCarrier().sendSpeakerRequest(flowId,
+                    new DumpGroupsForFlowHsRequest(protectedFlowPath.getYPointSwitchId()));
+        });
+
+
+    }
+
+    protected void receivedRules(State from, State to,
+                                 Event event, Object context) {
+        FlowDumpResponse flowDumpResponse = (FlowDumpResponse) context;
+        log.info("Switch rules received for switch {}", flowDumpResponse.getFlowSpeakerData().stream().findFirst()
+                .map(e -> e.getSwitchId().toString()).orElse("No switch Id"));
+        receivedRules.add(flowDumpResponse);
+        awaitingRules--;
```

Comment:
Thank you, please, elaborate on it, I see not concurrency problems here
receivedRules - updates only here, accessed strictly after update.
awaitingRules  - updates only here, there is no concurrent access

Reply:
It drew my attention because I thought that this class is not thread-safe. I asked just to make sure that concurrency has been taken into consideration. I see that nothing is overlooked here. So all is good. Thanks 👍